### DISCUSSION TITLE
Welcome to community Discussions!

### DISCUSSION THREAD

**Discussion by Relequestual** [json-schema-org/community#1] [General]
## 👋 Welcome!

This repositorie's Discussions is primerily a place for the community and the organisation teams to discuss community and organisational matters.

Please feel free to:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

Please do not use these Discussions to ask questions about working with JSON Schema itself.

We will likely have other discussion sections enabled soon, but for now you could always join our [active Slack server](https://json-schema.org/slack).

To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by GodwinOkwuibe** [comment]
Please I have the following question: 
When I try to install jsonschema using the default installation method (pin install jsonschema) on my ubuntu, it will install version 3.2.0. Unfortunately, this version is not compatible with the programme I want to run. If I try 'pip install jsonschema==4.2.1', it will throw the following error:

ERROR: Could not find a version that satisfies the requirement jsonschema==4.2.1 (from versions: 0.1a0, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8.0, 1.0.0, 1.1.0, 1.2.0, 1.3.0, 2.0.0, 2.1.0, 2.2.0, 2.3.0, 2.4.0, 2.5.0, 2.5.1, 2.6.0, 3.0.0a1, 3.0.0a2, 3.0.0a3, 3.0.0a4, 3.0.0a5, 3.0.0a6, 3.0.0b1, 3.0.0b2, 3.0.0b3, 3.0.0, 3.0.1, 3.0.2, 3.1.0, 3.1.1, 3.2.0, 4.0.0a1, 4.0.0a2, 4.0.0a3, 4.0.0a4, 4.0.0a5, 4.0.0a6, 4.0.0)
ERROR: No matching distribution found for jsonschema==4.2.1

Please can anyone help me out?

**Reply by karenetheridge** [reply]
You can find support information for that distribution on its pypi page: https://pypi.org/project/jsonschema/

**Comment by ghost** [comment]


**Reply by gregsdennis** [reply]
> I think it would supplant JSON as quickly as JSON is supplanting XML

After having a look at the readme and usage docs, I don't see how this can act as a data transfer mechanism, which is what XML and JSON are used for.

Can you expand on this?

If this is indeed the case, then, sure, I can see why it works be pertinent for us.  Thanks.

**Comment by ghost** [comment]
@gregsdennis Well, I don't see JSON as a data transfer mechanism either. I see JSON as a framework for deserializing data at transfer endpoints. Ribose would serve the same purpose in that context. The nature of the transfer channel is an orthogonal concern.

My comment, which you quote, was not intended as any sort of challenge, and I do appreciate the effort that JSON developers have put in to relieve the world of XML. I am using service APIs as a  demonstrative use case for ribose because it is a compelling but specialized use case for transduction. I throw a bit of shade on XML and JSON but, hey, it's nothing personal. I'm trying to help out legions of developers who have to absorb XML and successor frameworks ad infinitum or until they burn out. That is the best I could hope for and I have set my expectations appropriately. 

In the course of developing ribose I have gained a strong sense that service providers and consumers form a de facto collective working in a common domain and I suggest, out of the blue and for the sake of discussion, that they should be free to express the structure of information contained in messages exchanged within their service domain without restriction from external authorities (JSON and XML are external authorities in this sense). I don't think there is much value gained by enabling a Shopify server to parse and validate input describing the structure of a protein.  

Free expression of data representation within a service domain is entirely reasonable and appropriate as long as service providers and consumers have the tools to express and recognize concise and unambiguous syntax for the content encapsulated in their request and response APIs. WIth ginr+ribose they can do this rigorously and with (syntax, semantic)* maps in hand can very quickly obtain  ribose transducers and develop java targets to assimilate data received through the service APIs.  The only runtime dependency then is the ribose runtime, which is a very lightweight component. 

I think that ginr+ribose presents a very different perspective on computing and you will have to dig a little deeper into ribose if you want to understand what it is and how it can cover the same ground as XML and JSON while providing a more flexible and efficient way for developers to interact with streaming data. If you care to do so I suggest you look at [Everything is Hard](https://github.com/jrte/ribose/discussions/2) (for orientation) and then review [Ginr as a service](https://github.com/jrte/ribose/wiki#ginr-as-a-service-provider) for a concrete example. In the latter note the structure (pattern) of the input is entirely domain-specific. Also please note that the pattern describing the input is self-descriptive -- anyone with even basic knowledge of what is a finite-state automaton can look at it and infer its meaning at a glance. Have a look for yourself and tell me it ain't so ;). 

For some insight into CPU/RAM overhead see ribose [issue #19](https://github.com/jrte/ribose/issues/19), which I just fixed. I've included some runtime metrics (/usr/bin/time) and gc stats with links to before/after fix analyses of the respective gc logs. 

PS @karenetheridge I note that your are involved in control systems and you have probably seen many frameworks for building controllers. But you probably have never seen anything like ginr (If you have please let me know because I've scoured the net and found nothing like it). Subsequential binary patterns are a sweet way to express control systems.

**Reply by gregsdennis** [reply]
> I don't see JSON as a data transfer mechanism either. I see JSON as a framework for deserializing data at transfer endpoints.

What's the difference?  What you described is _exactly_ a "data transfer mechanism."  Data is serialized at one end (e.g. into XML/JSON), transferred, and deserialized at the other (e.g. from XML/JSON).

> ... they should be free to express the structure of information contained in messages exchanged within their service domain without **restriction from external authorities** (JSON and XML are external authorities in this sense).

How do data formats such as XML and JSON restrict data providers' communication with their consumers?  You make this claim, but you neglect to show why or provide examples.

Secondly, APIs _are_ free to use whatever data transfer they want.  They _choose_ JSON because it's standardized and there is almost guaranteed to be tooling for it on both the provider and consumer side, regardless of the language being used.  It makes the API easy to create and consume.

> I don't think there is much value gained by **enabling a Shopify server to parse and validate input describing the structure of a protein**.

These formats would never allow one API to (usefully) read and understand the data meant for another API.  The structure would be the same, but the data would be different.

In practice, if this were to happen, it would be expected that Shopify would respond with a 400 Bad Request because it doesn't understand the data that it received.

> Free expression of data representation within a service domain is entirely reasonable and appropriate **as long as service providers and consumers have the tools** to express and recognize concise and unambiguous syntax for the content encapsulated in their request and response APIs.

This is precisely the problem that XML and later JSON were designed to solve.  APIs that define their own data format must also develop tooling that consumers can use.  Doing all of that is a lot of work.  XML and JSON provide a standardized way to represent that data, which takes that burden off of the API.

Tooling for standardized formats can be generic and therefore agnostic and independent of the API, allowing reuse across the industry (which is what we have).

---

It sounds (to me) like you're trying to undo (redefine, maybe?) what the software industry has iteratively built over the past 30+ years.  Maybe this is just a lofty goal, but I don't think that there's anything for JSON Schema to do in this space for now.

**Reply by Relequestual** [reply]
Hey @jrte, it would be great if you could take this to a new Discussion to avoid clogging up this intro Discussion for the community =]

**Reply by ghost** [reply]
@Relequestual my apologies. Unless Greg objects you (or I, LMK if there are permission issues) can remove these posts and reclaim your space. No problem here.

I am just starting to trot ribose around to various communities to get feedback and perhaps some traction in moving it forward. Greg is correct with regard to his last point -- but I'm just an old fart with a keyboard. That said I see the "iterative" progression ASN.1 - XML - WSDL - JSON - ... and I think the series will eventually converge on something like ribose. It's just science fiction for now.

@gregsdennis If you or any of your colleagues want to continue this discussion privately or publicly please email jrte.project@gmail.com.

@ **all** thank you for your patience.

**Reply by Relequestual** [reply]
No problem, and thanks for sharing.

**Comment by jagpreetrahi** [comment]
Hii @Relequestual  I hope you are doing well , I am Jagpreet Singh from India Sub-continent , am a Backend developer , I am new here and wanna contribute here , so i just came across through  official website and i found some  issue regarding api calls and not able to load the page  after hits api , If you wanna see this issue ,I have some SS and i can share also , if you can assign me that issue which may be present currently . It's really appreciate to me.

**Comment by abdeldjalilhachimi** [comment]
Hello @Relequestual , I hope you are doing great! I am a Frontend web developer with years of experience, I have created a tool with React js  with Ajv to validate JSON schema I would like you to give me your opinion, and I am happy to join and share my experience and ideas 

Thanks in advance 

<img width="849" alt="Screenshot 2024-10-12 at 8 32 59 PM" src="https://github.com/user-attachments/assets/c9dea915-2286-4923-a668-9454b291606b">
